Adds MessageBar component to FluentUI

procload opened the pull request:
This pull request introduces a new `MessageBar` component to the Fluent Web Components. The `MessageBar` component is a notification bar that provides feedback about a user's actions or the state of an application. It can be used to display information, warnings, errors, or success messages.

Here are the key changes:

**New Component:**

* `MessageBar` component is introduced with various attributes including `shape`, `layout`, `intent`. It also provides a `dismissMessageBar` method to emit a `dismiss` event when the message bar is dismissed.

**Component Definition:**

* The `MessageBar` component is defined and registered with the Fluent Design System.
* The `MessageBar` component is exported with various options and styles.

**Component Options:**

* The `MessageBar` component options including `MessageBarLayout`, `MessageBarShape`, `MessageBarIntent`, and `MessageBarPoliteness` are defined.

**Component Styles:**

* The styles for the `MessageBar` component are defined. Different styles are applied based on the `shape`, `layout`, `intent`, and `politeness` attributes of the `MessageBar`.

**Component Template:**

* The template for the `MessageBar` component is created. The template includes slots for `actions` and `dismiss` and uses different icons based on the `intent` attribute.

**Component Integration Test:**

* An integration test for the `MessageBar` component is added. The test checks the default attributes, `dismiss` event, and changes to the `layout`, `intent`, and `shape` attributes.

**Component Story:**

* Storybook stories for the `MessageBar` component are added. The stories demonstrate the `MessageBar` with different shapes, layouts, intents, and politeness.

Closes #26709

fabricteam commented:
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

fabricteam commented on change/@fluentui-web-components-6860ddb3-4e1b-4679-9c94-1dfcf4bf40ca.json:
```diff
@@ -0,0 +1,7 @@
+{
```
## [🕵 fluentui-web-components-v3 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31708&commitId=500aa0f191afda3e6ad779223b8a0b7cd94c3604&env=prod&clientType=FLUENTUI)


  <details><summary><b>Accordion</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Accordion.Size - RTL.normal.chromium.png|2| Changed|
|Accordion.Size - Dark Mode.normal.chromium.png|2| Changed|
|Accordion.Size.normal.chromium.png|2| Changed|

</details>


  <details><summary><b>Avatar</b> 5 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar.Color - RTL.normal.chromium.png|1| Changed|
|Avatar.Color.normal.chromium.png|1| Changed|
|Avatar.Active.normal.chromium.png|5| Changed|
|Avatar.Active Appearance - Dark Mode.normal.chromium.png|2| Changed|
|Avatar.Active Appearance.normal.chromium.png|10| Changed|

</details>


  <details><summary><b>Badge</b> 6 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Badge.Shape.normal.chromium.png|1| Changed|
|Badge.Appearance - Dark Mode.normal.chromium.png|15| Changed|
|Badge.Size.normal.chromium.png|6| Changed|
|Badge.Default.normal.chromium.png|1| Changed|
|Badge.Appearance.normal.chromium.png|1| Changed|
|Badge.Color.normal.chromium.png|1| Changed|

</details>


  <details><summary><b>Button</b> 57 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Button.Large Icon Only - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large Icon Only.default.chromium.png|5| Changed|
|Button.Large Icon Only - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large With Icon - RTL.hover.chromium.png|4| Changed|
|Button.Large Icon Only.pressed.chromium.png|3| Changed|
|Button.Large With Icon - Dark Mode.default.chromium.png|4| Changed|
|Button.Large With Icon.default.chromium.png|5| Changed|
|Button.Large Icon Only.hover.chromium.png|4| Changed|
|Button.Large Icon Only - Dark Mode.default.chromium.png|4| Changed|
|Button.Large With Icon.hover.chromium.png|4| Changed|
|Button.Large With Icon - Dark Mode.hover.chromium.png|3| Changed|
|Button.Large With Icon - RTL.pressed.chromium.png|3| Changed|
|Button.Large With Icon - RTL.default.chromium.png|5| Changed|
|Button.Outline Disabled Focusable - Dark Mode.default.chromium.png|159| Changed|
|Button.Outline Disabled - Dark Mode.pressed.chromium.png|159| Changed|
|Button.Large With Icon - Dark Mode.pressed.chromium.png|5| Changed|
|Button.Large With Icon.pressed.chromium.png|3| Changed|
|Button.Outline Disabled - Dark Mode.hover.chromium.png|159| Changed|
|Button.Outline Disabled Focusable - Dark Mode.pressed.chromium.png|159| Changed|
|Button.Primary Disabled - Dark Mode.default.chromium.png|44| Changed|
|Button.Primary Disabled Focusable - Dark Mode.hover.chromium.png|44| Changed|
|Button.Primary Disabled Focusable.pressed.chromium.png|15| Changed|
|Button.Primary Disabled Focusable.default.chromium.png|15| Changed|
|Button.Outline Disabled - Dark Mode.default.chromium.png|159| Changed|
|Button.Primary Disabled Focusable - Dark Mode.default.chromium.png|44| Changed|
|Button.Primary Disabled - Dark Mode.pressed.chromium.png|44| Changed|
|Button.Primary Disabled.hover.chromium.png|15| Changed|
|Button.Primary Disabled Focusable - Dark Mode.pressed.chromium.png|44| Changed|
|Button.Outline Disabled Focusable - Dark Mode.hover.chromium.png|159| Changed|
|Button.Primary Disabled Focusable.hover.chromium.png|15| Changed|
|Button.Subtle Disabled Focusable.default.chromium.png|25| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.hover.chromium.png|187| Changed|
|Button.Subtle Disabled.default.chromium.png|25| Changed|
|Button.Subtle Disabled - Dark Mode.default.chromium.png|187| Changed|
|Button.Subtle Disabled.hover.chromium.png|25| Changed|
|Button.Primary Disabled - Dark Mode.hover.chromium.png|44| Changed|
|Button.Primary Disabled.pressed.chromium.png|15| Changed|
|Button.Subtle Disabled - Dark Mode.hover.chromium.png|187| Changed|
|Button.Subtle Disabled Focusable.pressed.chromium.png|25| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.pressed.chromium.png|187| Changed|
|Button.Primary Disabled.default.chromium.png|15| Changed|
|Button.Subtle Disabled Focusable - Dark Mode.default.chromium.png|187| Changed|
|Button.Subtle Disabled - Dark Mode.pressed.chromium.png|187| Changed|
|Button.Subtle Disabled Focusable.hover.chromium.png|25| Changed|
|Button.Transparent Disabled Focusable.default.chromium.png|27| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.pressed.chromium.png|199| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.default.chromium.png|199| Changed|
|Button.Transparent Disabled.pressed.chromium.png|27| Changed|
|Button.Transparent Disabled - Dark Mode.pressed.chromium.png|199| Changed|
|Button.Transparent Disabled - Dark Mode.hover.chromium.png|199| Changed|
|Button.Transparent Disabled - Dark Mode.default.chromium.png|199| Changed|
|Button.Transparent Disabled Focusable.pressed.chromium.png|27| Changed|
|Button.Subtle Disabled.pressed.chromium.png|25| Changed|
|Button.Transparent Disabled.hover.chromium.png|27| Changed|
|Button.Transparent Disabled.default.chromium.png|27| Changed|
|Button.Transparent Disabled Focusable - Dark Mode.hover.chromium.png|199| Changed|
|Button.Transparent Disabled Focusable.hover.chromium.png|27| Changed|

</details>


  <details><summary><b>Checkbox</b> 26 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Checkbox.Circular - Dark Mode.normal.chromium.png|792| Changed|
|Checkbox.Circular - RTL.normal.chromium.png|4733| Changed|
|Checkbox.Circular Disabled - Dark Mode.normal.chromium.png|760| Changed|
|Checkbox.Circular Disabled.normal.chromium.png|4711| Changed|
|Checkbox.Circular.normal.chromium.png|4734| Changed|
|Checkbox.Checked - Dark Mode.normal.chromium.png|982| Changed|
|Checkbox.Checked.normal.chromium.png|4855| Changed|
|Checkbox.Disabled - Dark Mode.normal.chromium.png|739| Changed|
|Checkbox.Default - Dark Mode.normal.chromium.png|814| Changed|
|Checkbox.Circular Disabled Checked - Dark Mode.normal.chromium.png|799| Changed|
|Checkbox.Default - RTL.normal.chromium.png|4751| Changed|
|Checkbox.Circular Checked.normal.chromium.png|4883| Changed|
|Checkbox.Disabled Checked.normal.chromium.png|4721| Changed|
|Checkbox.Default.normal.chromium.png|4751| Changed|
|Checkbox.Indeterminate - Dark Mode.normal.chromium.png|877| Changed|
|Checkbox.Indeterminate - Dark Mode.indeterminate.chromium.png|993| Changed|
|Checkbox.Indeterminate.indeterminate.chromium.png|4847| Changed|
|Checkbox.Disabled.normal.chromium.png|4705| Changed|
|Checkbox.Indeterminate.normal.chromium.png|4776| Changed|
|Checkbox.Circular Indeterminate.indeterminate.chromium.png|4869| Changed|
|Checkbox.Circular Indeterminate - Dark Mode.indeterminate.chromium.png|1102| Changed|
|Checkbox.Circular Checked - Dark Mode.normal.chromium.png|1080| Changed|
|Checkbox.Circular Indeterminate - Dark Mode.normal.chromium.png|936| Changed|
|Checkbox.Disabled Checked - Dark Mode.normal.chromium.png|777| Changed|
|Checkbox.Circular Indeterminate.normal.chromium.png|4786| Changed|
|Checkbox.Circular Disabled Checked.normal.chromium.png|4728| Changed|

</details>


  <details><summary><b>Divider</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Divider.With Svg.normal.chromium.png|3| Changed|
|Divider.With Svg - RTL.normal.chromium.png|3| Changed|
|Divider.Vertical With Svg.normal.chromium.png|4| Changed|

</details>


  <details><summary><b>Label</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Label.Required - RTL.normal.chromium.png|11| Changed|
|Label.Size Small.normal.chromium.png|661| Changed|

</details>


  <details><summary><b>Link</b> 10 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Link.Default.default.chromium.png|0| Added|
|Link.Default.hover.chromium.png|0| Added|
|Link.Default.pressed.chromium.png|0| Added|
|Link.Inline.default.chromium.png|0| Added|
|Link.Inline.hover.chromium.png|0| Added|
|Link.Inline.pressed.chromium.png|0| Added|
|Link.Subtle.default.chromium.png|0| Added|
|Link.Subtle.hover.chromium.png|0| Added|
|Link.Subtle.pressed.chromium.png|0| Added|
|Link.With Long Text.default.chromium.png|0| Added|

</details>


  <details><summary><b>MenuList</b> 16 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|MenuList.Checkbox With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox.2nd selected.chromium.png|1| Changed|
|MenuList.Checkbox With Icons.2nd selected.chromium.png|5| Changed|
|MenuList.Checkbox - RTL.2nd selected.chromium.png|1| Changed|
|MenuList.Checkbox With Icons.normal.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.2nd selected.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.normal.chromium.png|4| Changed|
|MenuList.Radio With Icons.1st selected.chromium.png|4| Changed|
|MenuList.Checkbox With Icons - RTL.2nd selected.chromium.png|5| Changed|
|MenuList.With Icons.hover menuitem.chromium.png|3| Changed|
|MenuList.Radio With Icons.2nd selected.chromium.png|4| Changed|
|MenuList.Radio With Icons.normal.chromium.png|4| Changed|
|MenuList.Radio With Icons - RTL.1st selected.chromium.png|4| Changed|
|MenuList.With Icons.normal.chromium.png|4| Changed|

</details>


  <details><summary><b>Slider</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Slider.Size Small.normal.chromium.png|1| Changed|
|Slider.Size Small.rightArrow.chromium.png|1| Changed|

</details>


  <details><summary><b>Switch</b> 34 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Switch.Checked.hover.chromium.png|5150| Changed|
|Switch.Checked - Dark Mode.hover.chromium.png|1169| Changed|
|Switch.Checked - Dark Mode.normal.chromium.png|1130| Changed|
|Switch.Default - Dark Mode.normal.chromium.png|1019| Changed|
|Switch.Default - Dark Mode.hover.chromium.png|1071| Changed|
|Switch.Default.hover.chromium.png|5044| Changed|
|Switch.Disabled - Dark Mode.hover.chromium.png|909| Changed|
|Switch.Checked - RTL.hover.chromium.png|5150| Changed|
|Switch.Disabled - RTL.normal.chromium.png|4914| Changed|
|Switch.Disabled - Dark Mode.normal.chromium.png|909| Changed|
|Switch.Disabled - RTL.hover.chromium.png|4914| Changed|
|Switch.Checked - RTL.normal.chromium.png|5132| Changed|
|Switch.Default - RTL.hover.chromium.png|5044| Changed|
|Switch.Disabled.hover.chromium.png|4914| Changed|
|Switch.Disabled Checked - Dark Mode.normal.chromium.png|1074| Changed|
|Switch.Disabled.normal.chromium.png|4914| Changed|
|Switch.Checked.normal.chromium.png|5132| Changed|
|Switch.Default - RTL.normal.chromium.png|5009| Changed|
|Switch.Disabled Checked.hover.chromium.png|5047| Changed|
|Switch.Disabled Checked - Dark Mode.hover.chromium.png|1074| Changed|
|Switch.Default.normal.chromium.png|5009| Changed|
|Switch.Disabled Checked.normal.chromium.png|5047| Changed|
|Switch.Label Above - RTL.hover.chromium.png|0| Removed|
|Switch.Label Above - RTL.normal.chromium.png|0| Removed|
|Switch.Label Above.hover.chromium.png|0| Removed|
|Switch.Label Above.normal.chromium.png|0| Removed|
|Switch.Label After - RTL.hover.chromium.png|0| Removed|
|Switch.Label After - RTL.normal.chromium.png|0| Removed|
|Switch.Label After.hover.chromium.png|0| Removed|
|Switch.Label After.normal.chromium.png|0| Removed|
|Switch.Label Before - RTL.hover.chromium.png|0| Removed|
|Switch.Label Before - RTL.normal.chromium.png|0| Removed|
|Switch.Label Before.hover.chromium.png|0| Removed|
|Switch.Label Before.normal.chromium.png|0| Removed|

</details>


  <details><summary><b>Text</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Text.Strikethrough.normal.chromium.png|236| Changed|
|Text.Block.normal.chromium.png|995| Changed|

</details>


  <details><summary><b>TextInput</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|TextInput.Size Large.normal.chromium.png|3| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_web-components"/>

procload commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,77 @@
+import { ElementViewTemplate, html, when } from '@microsoft/fast-element';
+import type { MessageBar } from './message-bar.js';
+
+const infoIcon = html`
```
Should SVGs be baked in, or slotted? Rethinking this now.

chrisdholt replied:
I don't think we should default these if there are designs given:
1. there are scenarios where you can _not_ have an icon at all, but the bundle will include all 4.
2. any scenario only requires *one* but you will get all 4

chrisdholt commented on packages/web-components/src/message-bar/message-bar.ts:
```diff
@@ -0,0 +1,70 @@
+   * HTML Attribute: `aria-labelledby`
+   */
+  @attr({ attribute: 'aria-labelledby' })
+  ariaLabelledBy: string | null = null;
```
I wouldn't observe any `aria-` attributes here as those come for free from the platform.

procload replied:
Can I just remove this, or do I need to setup anything else in the class to enable this?

chrisdholt replied:
This can just be removed, it can be set by implementors with:
```html
<message-bar aria-labelledby="foo"></message-bar>
```

chrisdholt commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,77 @@
+  return html<T>`
+    <template
+      role="status"
+      layout="${x => x.layout}"
+      shape="${x => x.shape}"
+      intent="${x => x.intent}"
```
If these are attributes they should reflect, no need to bind to the template

chrisdholt commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,77 @@
+export function messageBarTemplate<T extends MessageBar>(): ElementViewTemplate<T> {
+  return html<T>`
+    <template
+      role="status"
```
As we're moving to use `ElementInternals` can we move this to the class leveraging that and not sprout an attribute?

chrisdholt commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,77 @@
+      shape="${x => x.shape}"
+      intent="${x => x.intent}"
+      aria-live="${x => x.politeness}"
+      aria-labelledby="${x => x.ariaLabelledBy}"
```
As noted in the class, this can be removed

chrisdholt commented on packages/web-components/src/message-bar/message-bar.ts:
```diff
@@ -0,0 +1,70 @@
+   * @remarks
+   * HTML Attribute: `politeness`
+   */
+  @attr
+  politeness: MessageBarPoliteness = 'polite';
```
Status is a role which actually _is_ a live region defaulting the politeness. My bent here is to use documentation and/or an example to illustrate how you would override that rather than including a separate attribute. If I wanted to override the aria-live attribute, I think the easiest method would be:

```html
<message-bar aria-live="assertive"></message-bar>
```

procload replied:
Got it. I removed the `politeness` attribute and added a Assertive example to the Storybook.

chrisdholt commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,77 @@
+      aria-labelledby="${x => x.ariaLabelledBy}"
+    >
+      <span class="icon">
+        ${when(x => x.intent === 'success', successIcon)} ${when(x => x.intent === 'error', errorIcon)}
```
Yeah, I'd drop the icons as well because `when()` is going to bring a lot more size into this control

chrisdholt commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,77 @@
+        ${when(x => x.intent === 'success', successIcon)} ${when(x => x.intent === 'error', errorIcon)}
+        ${when(x => x.intent === 'warning', warningIcon)} ${when(x => x.intent === 'info', infoIcon)}
+      </span>
+      <div class="content">
+        <slot></slot>
+      </div>
+      <div class="actions">
+        <slot name="actions"></slot>
+      </div>
+      <div class="close">
+        <slot name="close"></slot>
+      </div>
```
I'm curious if we need all these wrapping divs or if `grid` can help setup "areas" where slots can be dropped without needing all this structure

procload replied:
I updated this to simplify the layout and removed a lot of the nesting. But I wasn't able to figure out how to do so with the default slot, so that one is still nested.

chrisdholt commented on packages/web-components/src/message-bar/message-bar.ts:
```diff
@@ -0,0 +1,70 @@
+  @attr({ attribute: 'aria-labelledby' })
+  ariaLabelledBy: string | null = null;
+
+  /**
+   * Sets the shape of the control.
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: `shape`
+   */
+  @attr
+  shape: MessageBarShape = 'rounded';
+
+  /**
+   * Sets the layout of the control.
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: `layout`
+   */
+  @attr
+  layout: MessageBarLayout = 'singleline';
+
+  /**
+   * Sets the intent of the control.
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: `intent`
+   */
+  @attr
+  intent: MessageBarIntent = 'info';
```
I'm going to be pushing a change to button/anchor button to use custom states with an attribute fallback - I think this is going to be "the way" given performance investigations showing no delta and in some cases, milliseconds of gain. Let's see if we can use those patterns here. I'll link the PR when it's up.

chrisdholt commented:
Can we add a benchmark for this? `.bench.ts` file to create an instance. In terms of the icons, I'd run one with the icons slotted (I believe what you've recently changed) and then I'd pull them out and run without and see what the delta is.

procload commented:
> Can we add a benchmark for this? `.bench.ts` file to create an instance. In terms of the icons, I'd run one with the icons slotted (I believe what you've recently changed) and then I'd pull them out and run without and see what the delta is.

How do I run the benchmarks? I created the message-bar.bench.ts file, but couldn't find documentation on how to get the results.

davatron5000 commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,84 @@
+          ${staticallyCompose(options.errorIcon)}
+        </span>
+      </span>
+      <span class="success">
+        <slot name="success-icon">
+          ${staticallyCompose(options.successIcon)}</span>
+        </slot>
+      </span>
```
Extra closing `</span>` here 

```suggestion
      <span class="success">
        <slot name="success-icon">
          ${staticallyCompose(options.successIcon)}
        </slot>
      </span>
```

procload commented:
@chrisdholt I ran the benchmarks with the SVGs in the template vs. outside of the template and added via a slot. Here's the results:

**SVGs in Template**
| Measurement       | Average    | Median     | Geometric Mean | Low         | High        |
|-------------------|------------|------------|----------------|-------------|-------------|
| javascriptTime    | 562.11     | 557.66     | 560.26         | 488.11      | 662.12      |
| layoutTime        | 1034.75    | 1128.78    | 949.44         | 576.23      | 2093.43     |
| memory            | 0          | 0          | 0              | 0           | 0           |
| domSize           | 2011       | 2011       | 2011.00        | 2011        | 2011        |
| lightDomSize      | 2011       | 2011       | 2011.00        | 2011        | 2011        |
| shadowDomSize     | 20000      | 20000      | 20000.00       | 20000       | 20000       |
| totalDomSize      | 22011      | 22011      | 22011.00       | 22011       | 22011       |
| maxLightBreadth   | 19         | 19         | 19.00          | 19          | 19          |
| maxLightDepth     | 15         | 15         | 15.00          | 15          | 15          |
| maxShadowBreadth  | 1          | 1          | 1.00           | 1           | 1           |
| maxShadowDepth    | 20         | 20         | 20.00          | 20          | 20          |

**SVGs outside of Template**
| Measurement       | Average    | Median     | Geometric Mean | Low         | High        |
|-------------------|------------|------------|----------------|-------------|-------------|
| javascriptTime    | 539.65     | 506.50     | 531.74         | 425.67      | 801.12      |
| layoutTime        | 1459.09    | 1269.74    | 1410.07        | 1123.07     | 2931.94     |
| memory            | 0          | 0          | 0              | 0           | 0           |
| domSize           | 2011       | 2011       | 2011.00        | 2011        | 2011        |
| lightDomSize      | 2011       | 2011       | 2011.00        | 2011        | 2011        |
| shadowDomSize     | 6000       | 6000       | 6000.00        | 6000        | 6000        |
| totalDomSize      | 8011       | 8011       | 8011.00        | 8011        | 8011        |
| maxLightBreadth   | 19         | 19         | 19.00          | 19          | 19          |
| maxLightDepth     | 15         | 15         | 15.00          | 15          | 15          |
| maxShadowBreadth  | 1          | 1          | 1.00           | 1           | 1           |
| maxShadowDepth    | 6          | 6          | 6.00           | 6           | 6           |

Given the performance benefits, I opted to **not** include the SVGs in the base template. You'll want to [review the message-bar.bench.ts file](https://github.com/microsoft/fluentui/pull/31708/files#diff-022af5ea507f6cd953badd168dae611e92c3c1519f702d1ec01c5704760ef91b) to ensure that I set this up correctly. But I was able to drastically simplify the markup in the template with this approach.

chrisdholt commented:
> @chrisdholt I ran the benchmarks with the SVGs in the template vs. outside of the template and added via a slot. Here's the results:
> 
> **SVGs in Template**
> 
> Measurement	Average	Median	Geometric Mean	Low	High
> javascriptTime	562.11	557.66	560.26	488.11	662.12
> layoutTime	1034.75	1128.78	949.44	576.23	2093.43
> memory	0	0	0	0	0
> domSize	2011	2011	2011.00	2011	2011
> lightDomSize	2011	2011	2011.00	2011	2011
> shadowDomSize	20000	20000	20000.00	20000	20000
> totalDomSize	22011	22011	22011.00	22011	22011
> maxLightBreadth	19	19	19.00	19	19
> maxLightDepth	15	15	15.00	15	15
> maxShadowBreadth	1	1	1.00	1	1
> maxShadowDepth	20	20	20.00	20	20
> **SVGs outside of Template**
> 
> Measurement	Average	Median	Geometric Mean	Low	High
> javascriptTime	539.65	506.50	531.74	425.67	801.12
> layoutTime	1459.09	1269.74	1410.07	1123.07	2931.94
> memory	0	0	0	0	0
> domSize	2011	2011	2011.00	2011	2011
> lightDomSize	2011	2011	2011.00	2011	2011
> shadowDomSize	6000	6000	6000.00	6000	6000
> totalDomSize	8011	8011	8011.00	8011	8011
> maxLightBreadth	19	19	19.00	19	19
> maxLightDepth	15	15	15.00	15	15
> maxShadowBreadth	1	1	1.00	1	1
> maxShadowDepth	6	6	6.00	6	6
> Given the performance benefits, I opted to **not** include the SVGs in the base template. You'll want to [review the message-bar.bench.ts file](https://github.com/microsoft/fluentui/pull/31708/files#diff-022af5ea507f6cd953badd168dae611e92c3c1519f702d1ec01c5704760ef91b) to ensure that I set this up correctly. But I was able to drastically simplify the markup in the template with this approach.

I'm going to pull this down and run a few. Those JS and layout numbers are massive.

procload commented:
@chrisdholt I just updated

> > @chrisdholt I ran the benchmarks with the SVGs in the template vs. outside of the template and added via a slot. Here's the results:
> > **SVGs in Template**
> > Measurement	Average	Median	Geometric Mean	Low	High
> > javascriptTime	562.11	557.66	560.26	488.11	662.12
> > layoutTime	1034.75	1128.78	949.44	576.23	2093.43
> > memory	0	0	0	0	0
> > domSize	2011	2011	2011.00	2011	2011
> > lightDomSize	2011	2011	2011.00	2011	2011
> > shadowDomSize	20000	20000	20000.00	20000	20000
> > totalDomSize	22011	22011	22011.00	22011	22011
> > maxLightBreadth	19	19	19.00	19	19
> > maxLightDepth	15	15	15.00	15	15
> > maxShadowBreadth	1	1	1.00	1	1
> > maxShadowDepth	20	20	20.00	20	20
> > **SVGs outside of Template**
> > Measurement	Average	Median	Geometric Mean	Low	High
> > javascriptTime	539.65	506.50	531.74	425.67	801.12
> > layoutTime	1459.09	1269.74	1410.07	1123.07	2931.94
> > memory	0	0	0	0	0
> > domSize	2011	2011	2011.00	2011	2011
> > lightDomSize	2011	2011	2011.00	2011	2011
> > shadowDomSize	6000	6000	6000.00	6000	6000
> > totalDomSize	8011	8011	8011.00	8011	8011
> > maxLightBreadth	19	19	19.00	19	19
> > maxLightDepth	15	15	15.00	15	15
> > maxShadowBreadth	1	1	1.00	1	1
> > maxShadowDepth	6	6	6.00	6	6
> > Given the performance benefits, I opted to **not** include the SVGs in the base template. You'll want to [review the message-bar.bench.ts file](https://github.com/microsoft/fluentui/pull/31708/files#diff-022af5ea507f6cd953badd168dae611e92c3c1519f702d1ec01c5704760ef91b) to ensure that I set this up correctly. But I was able to drastically simplify the markup in the template with this approach.
> 
> I'm going to pull this down and run a few. Those JS and layout numbers are massive.

I just updated the MessageBar benchmark template to include SVGs and actual content. Make sure to pull that down before you run any benchmarks.

procload commented:
@chrisdholt @davatron5000 Thanks for the feedback, guys! I implemented the changes requested and significantly simplified the component template. I also attempted to follow the state vs. attribute pattern that I see being refactored across the web components. Please let me know if I didn't implement anything correctly in this, but it all appears to be working :)

chrisdholt commented on packages/web-components/src/message-bar/message-bar.stories.ts:
```diff
@@ -0,0 +1,195 @@
+  <fluent-message-bar shape="${x => x.shape}" layout="${x => x.layout}" intent="${x => x.intent}">
+    <span slot="icon">${infoIcon}</span>
+    ${x => x.content}
+    <div class="actions" slot="actions">
+      <fluent-button size="small">Action</fluent-button>
+    </div>
```
Does this work if you just slot the button without the `<div>`?

chrisdholt commented on packages/web-components/src/message-bar/message-bar.stories.ts:
```diff
@@ -0,0 +1,195 @@
+  </fluent-message-bar>
+`);
+
+export const Assertive = renderComponent(html<MessageBarStoryArgs>`
```
Great addition to a story! Do we need to add `aria-labelledby` to the examples for a11y purposes? (Asking)

chrisdholt reviewed (commented):
Thanks @procload, looks pretty good - I see a few things no longer needed and had a couple of questions.

procload commented on packages/web-components/src/message-bar/message-bar.ts:
```diff
@@ -0,0 +1,87 @@
+   * HTML Attribute: `shape`
+   */
+  @attr
+  shape: MessageBarShape = 'rounded';
```
If we were to do this, does this mean to remove the `MessageBarShape` attribute from the class entirely and only use `shape="square"` to trigger the CSS?

procload commented:
@chrisdholt @radium-v @davatron5000 Thanks again for the feedback. I addressed all of the above. I had a few questions on if I implemented the below feedback correct. 

![image](https://github.com/microsoft/fluentui/assets/49389/882b4061-0436-4811-8be2-8ba940d58111)

And

![image](https://github.com/microsoft/fluentui/assets/49389/206be323-52e8-44bd-a882-76ded592f57c)

From what I could tell from inspecting the ProgressBar updates, it looks like I'm on the right track but wanted to double-check with you.

chrisdholt commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,25 @@
+    <div class="content">
+      <slot></slot>
+    </div>
+    <slot name="actions"></slot>
```
I dropped two actions in while testing this to see if it would support it, but it looks like some of the styling is off (spacing between the action and the close button, the two actions overlap in the same position):

<img width="1265" alt="image" src="https://github.com/microsoft/fluentui/assets/13071055/46c7664d-5d16-4d45-854d-d8ed24acea1f">

procload replied:
I wasn't able to figure out how to achieve the intended design without adding an additional wrapping `div` around the Actions. I didn't prescribe any spacing of the buttons within as I figured it would be best for the implementor to add this.

![image](https://github.com/microsoft/fluentui/assets/49389/40c49ad1-10b6-485c-a7a7-0aec07eba964)

chrisdholt commented on packages/web-components/src/message-bar/message-bar.styles.ts:
```diff
@@ -0,0 +1,126 @@
+    min-height: 36px;
+    align-items: center;
+    grid-template: 'icon body actions close' / auto 1fr auto auto;
+  }
```
Lets add CSS containment here - I totally missed this earlier in my review

procload replied:
Done!

procload commented on packages/web-components/src/message-bar/message-bar.template.ts:
```diff
@@ -0,0 +1,27 @@
+    <div class="actions">
+      <slot name="actions"></slot>
+    </div>
+    <slot name="close"></slot>
```
Great question. I was wrestling with this myself. I got `dismiss` from an example on the React docs. But I agree: We should be more consistent here.

@chrisdholt Do we have any naming standards that we're currently using across this library? Happy to conform to whatever is recommended.

chrisdholt replied:
At least within components we should stay aligned on a specific semantic. For the sake of _this_ component, it seems like `dismiss` is the action you'd take with an alert.

procload replied:
Great. Thanks for catching this. I renamed all `close` instances in the template and CSS to `dismiss`

chrisdholt commented on packages/web-components/src/message-bar/README.md:
```diff
@@ -15,37 +15,33 @@ The Fabric WC3 MessageBar extends `FASTElement`
-  <slot></slot>
-  <slot name="actions"></slot>
-  <slot name="close"></slot>
+<template>
```
You can remove the unused template here since you are no longer binding anything to it

procload replied:
Nice catch.

chrisdholt commented on packages/web-components/src/message-bar/message-bar.bench.ts:
```diff
@@ -0,0 +1,63 @@
+  messageBar.appendChild(document.createTextNode('message-bar'));
+
+  // Create and append icon slot
+  const icon = document.createElement('span');
+  icon.setAttribute('slot', 'icon');
+  icon.innerHTML = infoIcon;
```
You should be able to manipulate the SVG directly by setting the attribute and appending it (without innerHTML). With `innerHTML` to get this to render properly you'd have to include another directive.

chrisdholt commented on packages/web-components/src/message-bar/message-bar.bench.ts:
```diff
@@ -0,0 +1,63 @@
+  messageBar.appendChild(icon);
+
+  // Create and append content slot
+  const content = document.createElement('span');
+  content.setAttribute('slot', 'content');
+  content.appendChild(document.createTextNode('Accordion item'));
+  messageBar.appendChild(content);
```
Does this render correctly if you slot text only and not create a span? Curious for the entire implementation, no solely for the test - but if it does, we should drop the span from the test.

procload replied:
Got it. I think I did this correctly.

chrisdholt commented on packages/web-components/src/message-bar/message-bar.bench.ts:
```diff
@@ -0,0 +1,63 @@
+  messageBar.appendChild(content);
+
+  // Create and append dismiss slot
+  const dismiss = document.createElement('span');
+  dismiss.setAttribute('slot', 'dismiss');
+  dismiss.innerHTML = dismissed20Regular; // replace with your SVG content
```
Same comment here on modifying the SVG itself - if need be, just set the attributes on the SVG.

chrisdholt commented on packages/web-components/src/message-bar/message-bar.bench.ts:
```diff
@@ -0,0 +1,63 @@
+  messageBar.appendChild(dismiss);
+
+  // Create and append actions slot
+  const actions = document.createElement('button');
+  actions.setAttribute('slot', 'actions');
+  actions.textContent = '<fluent-button size="small">Action</fluent-button>';
```
This would be a button inside a button. Here you can create a fluent-button, set the size attribute and set the text.

procload replied:
Got it. I think I did this correctly.

procload commented:
@chrisdholt Looks like the benchmarks are now running locally. Is there any other feedback here that I can address?
